File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.py
import sys
input = sys.stdin.readline
N = int(input())
arr = list(map(int, input().split()))
dp = [1 for _ in range(N)]

for i in range(N):
    for j in range(i):
        if arr[i] > arr[j]:
            dp[i] = max(dp[i], dp[j]+1)

length = max(dp)
val = 0
for i in range(N):
    if dp[i] == length:
       val = arr[i]

ans_arr = []
for i in range(N-1, -1, -1):
    if dp[i] == length and arr[i] <= val:
        ans_arr.append(arr[i])
        length -= 1
        val = arr[i]


print(max(dp))
print(*ans_arr[::-1])

File: 백준/Gold/14002. 가장 긴 증가하는 부분 수열 4/가장 긴 증가하는 부분 수열 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	public static void main(String[] args) throws IOException {
		
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int [] arr = new int [N];
 		int[] dp = new int[N];
		StringTokenizer st = new StringTokenizer(br.readLine());
 		for(int i = 0; i < N; i++) {
			int n = Integer.parseInt(st.nextToken());
			arr[i] = n;
			dp[i] = 1;
		}
		for(int i = 0; i < N; i++) {
			for(int j = 0; j < i; j++) {
				if (arr[i] > arr[j]) {
					dp[i] = Math.max(dp[i], dp[j]+1);
				}
			}
		}
		int len = 0;
		int max_val = 0;
		for (int i = 0; i < N; i++) {
			if(len < dp[i]) {
				len = dp[i];
				max_val = arr[i];
			}
		}
		int len_ans = len;
		ArrayList<Integer> ans = new ArrayList<>();
		for (int i = N-1; i >= 0; i--) {
			if (dp[i] == len && arr[i] <= max_val) {
				ans.add(arr[i]);
				len -= 1;
				max_val = arr[i];
			}
		}
		System.out.println(len_ans);
		for (int i = ans.size()-1; i >= 0; i--) {
			System.out.print(ans.get(i)+" ");
		}
	}
}